File: src/main/java/leetcode/plan/graph/level1/LeetCode127.java
package leetcode.plan.graph.level1;

import java.util.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/10/20 22:50
 */
public class LeetCode127 {

    public static void main(String[] args) {
        System.out.println(new LeetCode127().
                ladderLength("ymain", "oecij", new ArrayList<>(Arrays.asList("ymann", "yycrj", "oecij", "ymcnj", "yzcrj", "yycij", "xecij", "yecij", "ymanj", "yzcnj", "ymain"))));
    }

    //"ymain"
    //"oecij"
    //["ymann","yycrj","oecij","ymcnj","yzcrj","yycij","xecij","yecij","ymanj","yzcnj","ymain"]

    public int ladderLength(String beginWord, String endWord, List<String> wordList) {

        HashSet<String> set = new HashSet<>(wordList);
        if (!set.contains(endWord)) {
            return 0;
        }

        Queue<String> queue = new LinkedList<>();
        queue.add(beginWord);
        HashSet<String> visited = new HashSet<>();
        visited.add(beginWord);
        int step = 1;
        while (!queue.isEmpty()) {
            step++;
            int size = queue.size();
            while (size > 0) {
                String cur = queue.poll();

                List<String> nexts = getList(cur, set);
                for (String next : nexts) {
                    if (next.equals(endWord)) {
                        return step;
                    }
                    if (visited.contains(next)) {
                        continue;
                    }
                    queue.add(next);
                    visited.add(next);
                }
                size--;
            }
        }
        return 0;
    }

    private List<String> getList(String cur, HashSet<String> set) {
        List<String> list = new ArrayList<>();
        char[] chars = cur.toCharArray();
        for (int i = 0; i < chars.length; i++) {
            char t = chars[i];
            for (char j = 'a'; j <= 'z'; j++) {
                if (j == t) {
                    continue;
                }
                chars[i] = j;
                String s = new String(chars);
                if (set.contains(s)) {
                    list.add(s);
                }
            }
            chars[i] = t;
        }
        return list;
    }
}

File: src/main/go/leetcode/lists/topinterview/LeetCode127.go
package topinterview

// "ymain"
// "oecij"
// ["ymann","yycrj","oecij","ymcnj","yzcrj","yycij","xecij","yecij","ymanj","yzcnj","ymain"]
func ladderLength(beginWord string, endWord string, wordList []string) int {
	q := make([]string, 0)
	q = append(q, beginWord)
	step := 0
	visited := make(map[string]bool)
	wordMp := make(map[string]bool)
	for i := 0; i < len(wordList); i++ {
		wordMp[wordList[i]] = true
	}
	if _, ok := wordMp[endWord]; !ok {
		return 0
	}
	visited[beginWord] = true
	for len(q) > 0 {
		size := len(q)
		for size > 0 {
			cur := q[0]
			q = q[1:]
			if cur == endWord {
				return step
			}
			nexts := getNext(cur, wordMp)
			//fmt.Print(cur)
			//fmt.Println(nexts)
			for _, next := range nexts {
				if visited[next] {
					continue
				}
				q = append(q, next)
				visited[next] = true
			}
			size--
		}
		step++
	}
	return 0
}

func getNext(cur string, wordMp map[string]bool) []string {
	list := make([]string, 0)
	mp := make(map[string]bool)
	for i := 0; i < len(cur); i++ {
		for ch := 'a'; ch <= 'z'; ch++ {
			t := cur[:i] + string(ch) + cur[i+1:]
			if wordMp[t] {
				mp[t] = true
			}
		}
	}
	for k := range mp {
		list = append(list, k)
	}
	return list
}
